File: Python3/895. Maximum Frequency Stack.py
class FreqStack:

    def __init__(self):
        self.frequencyCount = {}
        self.maxCount = 0
        self.stack = []

    def push(self, val: int) -> None:
        count = 1 + self.frequencyCount.get(val, 0)
        self.frequencyCount[val] = count
        self.maxCount = max(self.maxCount, count)
        while len(self.stack) < count:
            self.stack.append([])
        self.stack[count - 1].append(val)
    
    def pop(self) -> int:
        val = self.stack[self.maxCount - 1].pop()
        if len(self.stack[self.maxCount - 1]) == 0:
            self.maxCount -= 1
        self.frequencyCount[val] -= 1
        return val        


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()

File: C++/895. Maximum Frequency Stack.cpp
class FreqStack {
private:
    std::unordered_map<int, int> frequencyCount;
    std::vector<std::vector<int>> stack;
    int maxCount;
public:
    FreqStack() {
        maxCount = 0;
    }
    
    void push(int val) {
        int count = ++frequencyCount[val];
        while (stack.size() < count) {
            stack.push_back(std::vector<int>());
        }
        stack[count - 1].push_back(val);
        maxCount = max(count, maxCount);
    }
    
    int pop() {
        int val = stack[maxCount - 1].back();
        stack[maxCount - 1].pop_back();
        frequencyCount[val]--;
        if (stack[maxCount - 1].empty()) maxCount--;
        return val;
    }
};

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack* obj = new FreqStack();
 * obj->push(val);
 * int param_2 = obj->pop();
 */
